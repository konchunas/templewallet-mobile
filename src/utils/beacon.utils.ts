import { PartialTezosOperation, TezosOperationType } from '@airgap/beacon-sdk';

import { ParamsWithKind } from '../interfaces/op-params.interface';

export type SemiPartialTezosOperation = PartialTezosOperation & {
  fee?: string;
  gas_limit?: string;
  storage_limit?: string;
};

/* eslint-disable @typescript-eslint/no-explicit-any */
export const mapBeaconToTaquitoParams = (op: SemiPartialTezosOperation): ParamsWithKind => {
  const { fee, gas_limit, storage_limit, ...rest } = op;

  const walletParam = {
    ...rest,
    fee: fee as any,
    gasLimit: gas_limit as any,
    storageLimit: storage_limit as any
  };

  switch (walletParam.kind) {
    case TezosOperationType.ORIGINATION:
      const { script, ...orgRest } = walletParam;
      const { code, storage } = script as any; // Beacon has wrong types for script prop

      return {
        ...orgRest,
        kind: walletParam.kind as any, // Beacon and taquito has different enums for op.kind
        mutez: true, // The balance was already converted from Tez (ꜩ) to Mutez (uꜩ)
        init: storage,
        code
      };

    case TezosOperationType.TRANSACTION:
      const { destination, amount, parameters, ...txRest } = walletParam;

      return {
        ...txRest,
        kind: walletParam.kind as any, // Beacon and taquito has different enums for op.kind
        mutez: true, // The balance was already converted from Tez (ꜩ) to Mutez (uꜩ)
        amount: amount as any,
        to: destination,
        parameter: parameters
      };

    default:
      return walletParam as any; // Beacon and taquito has different enums for op.kind
  }
};
