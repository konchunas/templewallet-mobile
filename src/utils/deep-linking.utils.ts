import { Serializer } from '@airgap/beacon-sdk';
import { useEffect } from 'react';
import { Linking } from 'react-native';
import { URL } from 'react-native-url-polyfill';

import { BeaconHandler, isBeaconMessage } from '../beacon/beacon-handler';
import { isDefined } from './is-defined';

export const deepLinkHandler = async (url: string) => {
  try {
    const searchParams = new URL(url).searchParams;
    const type = searchParams.get('type');
    const data = searchParams.get('data');

    if (type === 'tzip10' && isDefined(data)) {
      const json = await new Serializer().deserialize(data);
      if (isBeaconMessage(json)) {
        await BeaconHandler.addPeer(json);
      }
    }
  } catch {}
};

export const useDeepLink = () => {
  useEffect(() => {
    const listener = ({ url }: { url: string }) => deepLinkHandler(url);

    Linking.addEventListener('url', listener);

    return () => Linking.removeEventListener('url', listener);
  }, []);
};
