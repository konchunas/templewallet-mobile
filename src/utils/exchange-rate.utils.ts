import { BigNumber } from 'bignumber.js';

import { isDefined } from './is-defined';

export const tokenToUsd = (amount: BigNumber.Value, exchangeRate?: number) => {
  if (isDefined(exchangeRate)) {
    return new BigNumber(amount).times(exchangeRate).decimalPlaces(2, BigNumber.ROUND_DOWN);
  }

  return new BigNumber(NaN);
};

export const usdToToken = (amount: BigNumber.Value, tokenDecimals: number, exchangeRate?: number) => {
  if (isDefined(exchangeRate) && exchangeRate !== 0) {
    return new BigNumber(amount).div(exchangeRate).decimalPlaces(tokenDecimals, BigNumber.ROUND_DOWN);
  }

  return new BigNumber(NaN);
};
