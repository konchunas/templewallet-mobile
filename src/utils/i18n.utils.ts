// TODO: replace all this functionality as soon as i18n that requires using hooks or store appears
export const makeRequiredErrorMessage = (fieldName: string) => `"${fieldName}" is a required field`;

const ordinalFormatter = new Intl.PluralRules('en-US', {
  type: 'ordinal'
});

const suffixes: Record<Intl.LDMLPluralRule, string> = {
  one: 'st',
  two: 'nd',
  few: 'rd',
  other: 'th',
  zero: 'th',
  many: 'th'
};

export const formatOrdinalNumber = (n: number) => `${n}${suffixes[ordinalFormatter.select(n)]}`;
