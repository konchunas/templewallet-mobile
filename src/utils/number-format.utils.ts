const ordinalFormatter = new Intl.PluralRules('en-US', {
  type: 'ordinal'
});

const suffixes: Record<Intl.LDMLPluralRule, string> = {
  one: 'st',
  two: 'nd',
  few: 'rd',
  other: 'th',
  zero: 'th',
  many: 'th'
};

export const formatOrdinalNumber = (n: number) => `${n}${suffixes[ordinalFormatter.select(n)]}`;
