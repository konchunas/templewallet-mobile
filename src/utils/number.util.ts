import { BigNumber } from 'bignumber.js';
import { isNaN } from 'lodash-es';

export const formatAssetAmount = (
  amount: BigNumber,
  roundingMode: BigNumber.RoundingMode = BigNumber.ROUND_DOWN,
  decimalPlace = 6
<<<<<<< HEAD
) => amount.decimalPlaces(amount.abs().lt(1000) ? decimalPlace : 2, roundingMode).toFixed();

export const invertSign = (amount: string) => new BigNumber(amount).times(-1).toString();

export const kFormatter = (num: number): string =>
  (Math.sign(num) * Math.round(Math.abs(num) / 1000)).toLocaleString() + ' K';
=======
) => {
  if (isNaN(amount.toNumber())) {
    return '';
  }

  return amount.decimalPlaces(amount.abs().lt(1000) ? decimalPlace : 2, roundingMode).toFixed();
};

export const roundFiat = (
  amount: BigNumber,
  roundingMode: BigNumber.RoundingMode = BigNumber.ROUND_DOWN
): BigNumber => {
  if (amount.isNaN()) {
    return new BigNumber(0);
  }

  return amount.decimalPlaces(2, roundingMode);
};

export const invertSign = (amount: string) => new BigNumber(amount).times(-1).toString();

export const kFormatter = (num: number): string | number => {
  if (isNaN(num)) {
    return '';
  }

  return (Math.sign(num) * Math.round(Math.abs(num) / 1000)).toLocaleString() + ' K';
};
>>>>>>> 4a8b805355b38151b514090b8ba56443d278f585
