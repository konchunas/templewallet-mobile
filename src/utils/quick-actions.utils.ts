import QuickActions, { ShortcutItem } from 'react-native-quick-actions';

import { EmptyFn } from '../config/general';

export const initQuickActions = () =>
  QuickActions.setShortcutItems([
    {
      type: 'Hide balance',
      title: 'Hide balance',
      icon: 'Prohibit',
      userInfo: {
        url: ''
      }
    }
  ]);

export interface QuickActionParams {
  data: ShortcutItem;
  isBalanceHiddenState: boolean;
  hideBalanceHandler: EmptyFn;
}

export const quickActionHandler = ({ data, isBalanceHiddenState, hideBalanceHandler }: QuickActionParams) => {
  // TODO: find a solution to handle quick actions on cold start
  if (data && !isBalanceHiddenState) {
    hideBalanceHandler();
  }
};
