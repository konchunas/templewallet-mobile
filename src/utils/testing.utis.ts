import { PartialObserver } from 'rxjs';

import { EventFn } from '../config/general';

export const rxJsTestingHelper = <T>(callback: EventFn<T>, done: jest.DoneCallback): PartialObserver<T> => ({
  next: data => {
    try {
      callback(data);

      done();
    } catch (e) {
      done(e);
    }
  },
  error: e => done(e)
});
