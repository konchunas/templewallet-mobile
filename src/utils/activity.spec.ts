import { ActivityStatusEnum } from '../enums/activity-status.enum';
import { ActivityTypeEnum } from '../enums/activity-type.enum';
import { ActivityInterface } from '../interfaces/activity.interface';
import { groupActivitiesByHash } from './activity.utils';

describe('groupActivitiesByHash', () => {
  const mockMember = {
    address: 'address2',
    alias: 'alias2'
  };

  const operations: ActivityInterface[] = [
    {
      type: ActivityTypeEnum.Transaction,
      status: ActivityStatusEnum.Applied,
      hash: 'hash1',
      amount: '0',
      address: 'address1',
      timestamp: 1,
      entrypoint: 'transfer',
      source: mockMember,
      destination: {
        address: 'address1',
        alias: 'alias1'
      }
    },
    {
      type: ActivityTypeEnum.Transaction,
      status: ActivityStatusEnum.Pending,
      hash: 'hash2',
      amount: '1',
      timestamp: 10,
      entrypoint: '',
      source: mockMember,
      destination: {
        address: 'address3',
        alias: 'alias3'
      }
    },
    {
      type: ActivityTypeEnum.Delegation,
      status: ActivityStatusEnum.Applied,
      hash: 'hash3',
      amount: '0',
      timestamp: 6,
      source: mockMember,
      destination: {
        address: 'address3',
        alias: 'alias3'
      }
    }
  ];

  it('should sort operations and transfers by timestamp in descending order', () => {
    const transfers: ActivityInterface[] = [
      {
        type: ActivityTypeEnum.Transaction,
        status: ActivityStatusEnum.Applied,
        hash: 'hash4',
        amount: '0',
        timestamp: 8,
        entrypoint: 'transfer',
        source: {
          address: 'address4',
          alias: 'alias4'
        },
        destination: mockMember
      }
    ];

    expect(groupActivitiesByHash(operations, transfers)).toMatchSnapshot();
  });

  it('should push operations and transfers with the same hash into one group and not mutate activities', () => {
    const transfers: ActivityInterface[] = [
      {
        type: ActivityTypeEnum.Transaction,
        status: ActivityStatusEnum.Pending,
        hash: 'hash2',
        amount: '1',
        timestamp: 9,
        entrypoint: '',
        source: mockMember,
        destination: {
          address: 'address4',
          alias: 'alias4'
        }
      },
      {
        type: ActivityTypeEnum.Transaction,
        status: ActivityStatusEnum.Applied,
        hash: 'hash4',
        amount: '0',
        timestamp: 8,
        entrypoint: 'transfer',
        source: {
          address: 'address4',
          alias: 'alias4'
        },
        destination: mockMember
      }
    ];

    expect(groupActivitiesByHash(operations, transfers)).toMatchSnapshot();
  });
});
