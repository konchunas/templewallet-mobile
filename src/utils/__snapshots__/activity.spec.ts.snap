// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`activity should push operations and transfers with the same hash into one group and not mutate activities 1`] = `
Array [
  Array [
    Object {
      "amount": "1",
      "destination": Object {
        "address": "address3",
        "alias": "alias3",
      },
      "entrypoint": "",
      "hash": "hash2",
      "source": Object {
        "address": "address2",
        "alias": "alias2",
      },
      "status": "pending",
      "timestamp": 10,
      "type": "transaction",
    },
    Object {
      "amount": "1",
      "destination": Object {
        "address": "address4",
        "alias": "alias4",
      },
      "entrypoint": "",
      "hash": "hash2",
      "source": Object {
        "address": "address2",
        "alias": "alias2",
      },
      "status": "pending",
      "timestamp": 9,
      "type": "transaction",
    },
  ],
  Array [
    Object {
      "amount": "0",
      "destination": Object {
        "address": "address2",
        "alias": "alias2",
      },
      "entrypoint": "transfer",
      "hash": "hash4",
      "source": Object {
        "address": "address4",
        "alias": "alias4",
      },
      "status": "applied",
      "timestamp": 8,
      "type": "transaction",
    },
  ],
  Array [
    Object {
      "amount": "0",
      "destination": Object {
        "address": "address3",
        "alias": "alias3",
      },
      "hash": "hash3",
      "source": Object {
        "address": "address2",
        "alias": "alias2",
      },
      "status": "applied",
      "timestamp": 6,
      "type": "delegation",
    },
  ],
  Array [
    Object {
      "address": "address1",
      "amount": "0",
      "destination": Object {
        "address": "address1",
        "alias": "alias1",
      },
      "entrypoint": "transfer",
      "hash": "hash1",
      "source": Object {
        "address": "address2",
        "alias": "alias2",
      },
      "status": "applied",
      "timestamp": 1,
      "type": "transaction",
    },
  ],
]
`;

exports[`activity should sort operations and transfers by timestamp in descending order 1`] = `
Array [
  Array [
    Object {
      "amount": "1",
      "destination": Object {
        "address": "address3",
        "alias": "alias3",
      },
      "entrypoint": "",
      "hash": "hash2",
      "source": Object {
        "address": "address2",
        "alias": "alias2",
      },
      "status": "pending",
      "timestamp": 10,
      "type": "transaction",
    },
  ],
  Array [
    Object {
      "amount": "0",
      "destination": Object {
        "address": "address2",
        "alias": "alias2",
      },
      "entrypoint": "transfer",
      "hash": "hash4",
      "source": Object {
        "address": "address4",
        "alias": "alias4",
      },
      "status": "applied",
      "timestamp": 8,
      "type": "transaction",
    },
  ],
  Array [
    Object {
      "amount": "0",
      "destination": Object {
        "address": "address3",
        "alias": "alias3",
      },
      "hash": "hash3",
      "source": Object {
        "address": "address2",
        "alias": "alias2",
      },
      "status": "applied",
      "timestamp": 6,
      "type": "delegation",
    },
  ],
  Array [
    Object {
      "address": "address1",
      "amount": "0",
      "destination": Object {
        "address": "address1",
        "alias": "alias1",
      },
      "entrypoint": "transfer",
      "hash": "hash1",
      "source": Object {
        "address": "address2",
        "alias": "alias2",
      },
      "status": "applied",
      "timestamp": 1,
      "type": "transaction",
    },
  ],
]
`;
